flax/nnx/graph.py

```diff
@@ -2885,6 +2885,45 @@ def pure_caller(accessor: DelayedAccessor, *args, **kwargs):
   return CallableProxy(pure_caller)  # type: ignore
 
 
+def set_metadata(node: A, /, *, only: filterlib.Filter = Variable, **metadata: tp.Mapping[str, tp.Any]) -> A:
+  """Sets the metadata of all :class:`Variable` objects in the given graph node.
+
+  Example::
+
+    >>> from flax import nnx
+    >>> import jax, jax.numpy as jnp
+    ...
+    >>> class Foo(nnx.Module):
+    ...   def __init__(self):
+    ...     self.param = nnx.Param(0.0)
+    ...     self.variable = nnx.Variable(0.0)
+    ...
+    >>> node = Foo()
+    ...
+    >>> # set differentiable to False for all nnx.Param objects
+    >>> node = nnx.graph.set_metadata(node, differentiable=False, only=nnx.Param)
+    ...
+    >>> # check that only the nnx.Param was updated
+    >>> assert node.param.get_metadata('differentiable') is False
+
+  Args:
+    node: A graph node object.
+    only: A Filter to specify which :class:`Variable` objects to set metadata for.
+    metadata: Key-value pairs to set as metadata on the :class:`Variable` objects.
+  Returns:
+    The graph node with updated :class:`Variable` metadata.
+  """
```

Comment:
Because the updates are being done in-place via `set_metadata` there is no need to return a new node. You can change `split(node, only, ...)` with `variable_state = state(node, only)`.